### ISSUE TITLE
Audio via HomeKit?

### ISSUE THREAD

**Issue by lightbody** [briis/unifiprotect#193]
Hey there -

Based on previously closed issues, it looks like audio in Home Assistant simply doesn't work. I'm assuming that is still the case, but please let me know if I'm wrong.

What I'm interested in is (maybe) slightly different: I'd like to have audio come through when I say "hey siri show me the front door" and get a camera feed that includes audio. Why siri vs the Unifi Protect app? It's faster and more convenient and works on our Apple TV.

And now that I took the time to move each camera to it's own dedicated HomeKit integration instance running in "accessory mode" and configured each camera to send the raw H264 stream, the latency is near-zero. Before it was just as bad as it still is in Home Assistant: 3-7 seconds.

So now that I have really good picture + low latency HomeKit, I really want audio too :) Is the limitation re: HA + audio also impacting this? I was hoping that because I configured them to have native H264 w/ no transcoding, it might mean the RTSP stream was going straight to my phone, audio and all.

Thanks!

**Comment by klausdm** [comment]
Hi. I have audio in Home Assistant on all my cameras. Have not yet tried HomeKit integration tho.

**Comment by lightbody** [comment]
> Hi. I have audio in Home Assistant on all my cameras. Have not yet tried HomeKit integration tho.

Intresting. I was looking at issue #87 and it sounded like that just wasn't possible right now. But perhaps that changed since last year. Anything special you can recall you're doing to get audio working? You're getting audio through your browser when viewing the normal life feed / camera lovelace card?

**Comment by klausdm** [comment]
> Anything special you can recall you're doing to get audio working? You're getting audio through your browser when viewing the normal life feed / camera lovelace card?
Nope. Just standard setup. I have a UDMP and G3, G3 Dome and G3 flex cams. All have sound in PC browser and Home Assistant iOS app.

**Comment by lightbody** [comment]
Very odd. I'm running everything in Docker containers, but it's on a somewhat dated OS (Ubuntu 16.04)... I wonder if somehow some old libraries are finding their way into my stack and messing things up.

**Comment by klausdm** [comment]
Not sure. I am running HASS.io on a NUC....

**Comment by klausdm** [comment]
Which revision core are you running. Audio through the stream component was released somewhat late last year. https://github.com/home-assistant/core/pull/38846

**Comment by lightbody** [comment]
Ah, good tip. I've noticed that my audio process seems to be not running... I'll poke around here.

![image](https://user-images.githubusercontent.com/50803/107163195-1a5b5400-695d-11eb-9751-68f053f1dd32.png)

**Comment by briis** [comment]
As @klausdm points out, audio is now working with the `stream` component. I cannot recall the exact version of HA when this was launched, but it was after issue #83 was closed. I have also not played around with the Homekit Integration, so I have no experience to share here.

I could though be interested in understanding exactly how you have set it up. Did you use the Homebridge integration for UniFi Protect, or how did you add the Cameras to Homekit?

**Comment by klausdm** [comment]
I am interested in the HomeKit integration you’ve done as well :-)

**Comment by briis** [comment]
I did not do the Homebridge Integration, this is done by @hjdhjd Have a look here: https://github.com/hjdhjd/homebridge-unifi-protect

**Comment by bdraco** [comment]
If you are using Home Assistant to get the camera to `HomeKit` you need to enable this 

https://www.home-assistant.io/integrations/homekit/#support_audio

**Comment by lightbody** [comment]
@bdraco Thank you for that tip - I totally missed that. Do you by chance know how I would go about configuring that for the HomeKit integrations that I configured via the GUI (and are setup in accessory mode rather than bridge mode)?

@briis @klausdm My HomeKit setup is pretty straightforward, I think. I have one primary integration setup via `configuration.yaml` that merges in the various domains and individual entities that don't already have native HomeKit integration. I do that because I figure talking from iPhone -> Apple TV -> Lutron Caseta is more reliable than from iPhone -> Apple TV -> Home Assistant -> Lutron Caseta, but I might revisit this decision someday. 

So the config looks roughly like:

```yaml
homekit:
  auto_start: False
  filter:
    include_domains:
      - cover
      - lock
#      - camera
      - sensor
      - binary_sensor
    include_entities:
      - switch.master_bathroom_fan
      - switch.powder_room_fan
    exclude_entities:
      - binary_sensor.front_door_open
      - lock.front_door
      - binary_sensor.updater
      - cover.playroom_shade_switch
      - cover.master_bedroom_shade_switch
    exclude_entity_globs:
      - sensor.motion_recording_*
      - binary_sensor.motion_*
```

I used to have my cameras configure in here, but that had the terrible ~10s lag. Now I separately have each camera wired up through additional HomeKit integrations, which I set up using the point-and-click interface:

<img width="285" alt="image" src="https://user-images.githubusercontent.com/50803/108269892-7d9a7280-7123-11eb-8c10-df473c88d5ad.png">

<img width="403" alt="image" src="https://user-images.githubusercontent.com/50803/108269933-8c812500-7123-11eb-940a-1dfa3047e9d7.png">

That results in the cameras showing up in HomeKit like I had before, but without the horrible lag. Now I just have to figure out audio! I'm hopeful that @bdraco's tip will work for me (once I figure out how to set that parameter). But I'm also not optimistic because I can't get audio out of normal HASS right now, partly due to my funky non-standard Docker setup:

<img width="917" alt="image" src="https://user-images.githubusercontent.com/50803/108270103-d10cc080-7123-11eb-9112-7e0184dda317.png">

With the `hassio_audio` not working anymore (it used to a few months back), the logs reporting:

```
[s6-init] making user provided files available at /var/run/s6/etc...exited 0.
[s6-init] ensuring user provided files have correct perms...exited 0.
[fix-attrs.d] applying ownership & permissions fixes...
[fix-attrs.d] scripts: applying... 
[fix-attrs.d] scripts: exited 0.
[fix-attrs.d] done.
[cont-init.d] executing container initialization scripts...
[cont-init.d] alsa-mixer.sh: executing... 
[13:26:17] INFO: Adjust ALSA mixer settings for /dev/snd/controlC0
[13:26:18] INFO: Adjust ALSA mixer settings for /dev/snd/controlC1
[cont-init.d] alsa-mixer.sh: exited 0.
[cont-init.d] filesystem.sh: executing... 
[cont-init.d] filesystem.sh: exited 0.
[cont-init.d] pulse-config.sh: executing... 
[cont-init.d] pulse-config.sh: exited 0.
[cont-init.d] done.
[services.d] starting services
[services.d] done.
W: [pulseaudio] main.c: Running in system mode, but --disallow-module-loading not set.
D: [pulseaudio] core-rtclock.c: Timer slack is set to 50 us.
D: [pulseaudio] core-util.c: setpriority() worked.
I: [pulseaudio] core-util.c: Successfully gained nice level -11.
I: [pulseaudio] main.c: Found user 'root' (UID 0) and group 'root' (GID 0).
W: [pulseaudio] main.c: Home directory of user 'root' is not '/var/run/pulse', ignoring.
W: [pulseaudio] caps.c: Normally all extra capabilities would be dropped now, but that's impossible because PulseAudio was built without capabilities support.
I: [pulseaudio] main.c: Successfully changed user to "root".
I: [pulseaudio] main.c: This is PulseAudio 14.2
D: [pulseaudio] main.c: Compilation CFLAGS: Not yet supported on meson
D: [pulseaudio] main.c: Running on host: Linux x86_64 4.4.0-142-generic #168-Ubuntu SMP Wed Jan 16 21:00:45 UTC 2019
D: [pulseaudio] main.c: Found 4 CPUs.
I: [pulseaudio] main.c: Page size is 4096 bytes
D: [pulseaudio] main.c: Compiled with Valgrind support: no
D: [pulseaudio] main.c: Running in valgrind mode: no
D: [pulseaudio] main.c: Running in VM: no
D: [pulseaudio] main.c: Running from build tree: no
D: [pulseaudio] main.c: Optimized build: yes
D: [pulseaudio] main.c: All asserts enabled.
I: [pulseaudio] main.c: Machine ID is hassio-audio.
I: [pulseaudio] main.c: Using runtime directory /var/run/pulse.
E: [pulseaudio] core-util.c: Failed to create secure directory (/data/states): Operation not permitted
[cont-finish.d] executing container finish scripts...
[cont-finish.d] done.
[s6-finish] waiting for services.
[s6-finish] sending all processes the TERM signal.
[s6-finish] sending all processes the KILL signal and exiting.
```

Anyway, that is pretty much it! Any advice or questions? :)

**Comment by bdraco** [comment]
There isn't any way to configure audio via the UI at this time.

I could probably add support for it to HomeKit options but there the HomeKit reviews are backlogged and there are a few that would block that including https://github.com/home-assistant/core/pull/46473

**Comment by briis** [comment]
I am not a Homekit expert, so how do you do this? "Now I separately have each camera wired up through additional HomeKit integrations" I assume they are all UniFi Cameras?

**Comment by bdraco** [comment]
The cameras are being created in accessory mode (https://github.com/home-assistant/core/pull/46473 will do this automatically for new users once it merges)

In accessory mode, the camera gets it own encrypted homekit connection.  In bridge mode it has to share with everything else on the bridge which can be slow.

**Comment by lightbody** [comment]
@briis Yes, all my cameras are Unifi cameras. They were all originally running through a single HomeKit configuration (see the commented out "cameras" entity in my config above). But I noticed some logs the other day that recommended each camera get it's own homekit integration in "accessory" mode, which I did via point-and-click configuration, and that resulted in much better streaming performance for my 10 cameras. LMK if you'd like a step-by-step, it's the least I could do considering how much I've benefited from your work with this plugin.

@bdraco Thank you for the tip re: `.storage/core.config_entries`! That worked like a charm and now I have streaming audio in when I say "hey siri show me the front door"! It's a bit choppy, but frankly it is in the Unifi native app too unless I pause the stream and give it a few seconds to buffer. 

Now all I have to figure out is how to get audio to stream out of the normal Lovelace cards in HA. It's unclear if it's related to the failed module on my system: I can't tell what the hassio_audio component is even for (it _seems_ like it is for playing audio out of the host device speakers and not streaming audio).

**Comment by georgeboot** [comment]
I was able to enable audio in my Homekit camera stream of my doorbell. Thanks for the info!

Something I noticed that the audio is stuttering all the time. The sound comes for 2 seconds and than goes for 2 seconds all the time. I've tested on multiple devices.

I initially assumed the audio must be transcoded and my Pi (hassos) isn't fast enough, but the load on the pi doesn't show any spikes.

Any suggestions where this might be coming from?

**Comment by briis** [comment]
No unfortunately, this is not my area of expertise. In the forum there is another user reporting the same https://community.home-assistant.io/t/custom-component-unifi-protect/158041/625
